File: app/models/discount.py
from .db import db



class Discount(db.Model):
    __tablename__ = 'discounts'

    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(300), nullable=False, unique=True)
    website = db.Column(db.String(300), nullable=False)
    instagram = db.Column(db.String(100))
    country = db.Column(db.String(4), nullable=False)

    is_yarn = db.Column(db.Boolean, nullable=False)
    is_fiber = db.Column(db.Boolean, nullable=False)
    is_notions = db.Column(db.Boolean, nullable=False)

    is_pattern = db.Column(db.Boolean, nullable=False)
    is_knitting = db.Column(db.Boolean, nullable=False)
    is_crochet = db.Column(db.Boolean, nullable=False)
    is_embroidery = db.Column(db.Boolean, nullable=False)

    discount_code = db.Column(db.String(100), nullable=False)
    discount_details = db.Column(db.Text, nullable=False)

    def to_dict(self):
        return {
            'id': self.id,
            'name': self.name,
            'website': self.website,
            'instagram': self.instagram,
            'country': self.country,
            'is_yarn': self.is_yarn,
            'is_fiber': self.is_fiber,
            'is_notions': self.is_notions,
            'is_pattern': self.is_pattern,
            'is_knitting': self.is_knitting,
            'is_crochet': self.is_crochet,
            'is_embroidery': self.is_embroidery,
            'discount_code': self.discount_code,
            'discount_details': self.discount_details
        }

    def to_JSON(self):
        return {
            'id': self.id,
            'name': self.name,
            'website': self.website,
            'instagram': self.instagram,
            'country': self.country,
            'isYarn': self.is_yarn,
            'isFiber': self.is_fiber,
            'isNotions': self.is_notions,
            'isPattern': self.is_pattern,
            'isKnitting': self.is_knitting,
            'isCrochet': self.is_crochet,
            'isEmbroidery': self.is_embroidery,
            'discountCode': self.discount_code,
            'discountDetails': self.discount_details
        }

File: react-app/src/store/discount.js
import { csrfFetch } from "../helpers";

const LOAD_DISCOUNTS = 'discounts/loadDiscounts';
const CREATE_DISCOUNT = 'discounts/createDiscount';
const EDIT_DISCOUNT = 'discounts/editDiscount';
const DELETE_DISCOUNT = 'discounts/deleteDiscount';

export const loadDiscounts = () => async (dispatch) => {
    const response = await csrfFetch('/api/discounts/');
    const allDiscounts = await response.json();
    dispatch(loadDiscountsAction(allDiscounts));
    return allDiscounts;
};

const loadDiscountsAction = (allDiscounts) => ({
    type: LOAD_DISCOUNTS,
    allDiscounts
});

export const createDiscount = ({
    name,
    website,
    instagram,
    country,
    isYarn,
    isFiber,
    isNotions,
    isPattern,
    isKnitting,
    isCrochet,
    isEmbroidery,
    discountCode,
    discountDetails
}) => async (dispatch) => {
    const response = await csrfFetch('/api/discounts/', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            name,
            website,
            instagram,
            country,
            is_yarn: isYarn,
            is_fiber: isFiber,
            is_notions: isNotions,
            is_pattern: isPattern,
            is_knitting: isKnitting,
            is_crochet: isCrochet,
            is_embroidery: isEmbroidery,
            discount_code: discountCode,
            discount_details: discountDetails
        })
    });

    if (response.ok) {
        const discount = await response.json();
        dispatch(createDiscountAction(discount));
        return discount;
    };
};

const createDiscountAction = (discount) => ({
    type: CREATE_DISCOUNT,
    discount
});

export const editDiscount = ({
    id,
    name,
    website,
    instagram,
    country,
    isYarn,
    isFiber,
    isNotions,
    isPattern,
    isKnitting,
    isCrochet,
    isEmbroidery,
    discountCode,
    discountDetails
}) => async (dispatch) => {
    const response = await csrfFetch('/api/discounts/', {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            id,
            name,
            website,
            instagram,
            country,
            is_yarn: isYarn,
            is_fiber: isFiber,
            is_notions: isNotions,
            is_pattern: isPattern,
            is_knitting: isKnitting,
            is_crochet: isCrochet,
            is_embroidery: isEmbroidery,
            discount_code: discountCode,
            discount_details: discountDetails
        })
    });

    if (response.ok) {
        const discount = await response.json();
        dispatch(editDiscountAction(discount));
        return discount;
    };
};

const editDiscountAction = (discount) => ({
    type: EDIT_DISCOUNT,
    discount
});

export const deleteDiscount = (discountId) => async (dispatch) => {
    const response = await csrfFetch('/api/discounts/', {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ id: discountId })
    });
    const discount = await response.json();
    dispatch(deleteDiscountAction(discount, discountId));
};

const deleteDiscountAction = (discount, discountId) => {
    return {
        type: DELETE_DISCOUNT,
        discount,
        discountId
    };
};

const discountReducer = (state = {}, action) => {
    let newState = { ...state };

    switch (action.type) {
        case LOAD_DISCOUNTS:
            action.allDiscounts.forEach(discount => {
                newState[discount.id] = discount;
            });
            return newState;
        case CREATE_DISCOUNT:
        case EDIT_DISCOUNT:
            newState[action.discount.id] = action.discount;
            return newState;
        case DELETE_DISCOUNT:
            delete newState[action.discountId];
            return newState;

        default:
            return state;
    };
};

export default discountReducer;
